iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -0,0 +1,214 @@
+
+    tracing::debug!("done {} {}", work.src_file, output.status);
+
+    Ok(())
+}
+
+async fn do_rsync(work_list: &Vec<Work>) -> Result<Vec<FidError>, ImlAgentError> {
+    let result = Vec::new();
+
+    stream::iter(work_list)
+        .map(|xs| Ok::<_, std::io::Error>(xs))
+        .try_for_each_concurrent(100, |fids| {
+            actually_rsync(fids)
+                .or_else(|e| {
+                    println!("error: {}", e);
+                    future::ok(())
+                })
+                .map_ok(|_| println!("done - actually"))
+        })
+        .await?;
+
+    Ok(result)
+}
+
+async fn archive_fids(
+    llapi: LlapiFid,
+    dest_src: &String,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = Vec::new();
+
+    let mut rsync_list: Vec<Work> = Vec::new();
+
+    for fid in fid_list {
+        let res = llapi.fid2path(&fid.fid);
+
+        if res.is_err() {
+            tracing::error!("llapi: failed on fid {}", &fid.fid);
+            continue;
+        }
+
+        let fid_path = res.ok().unwrap();
+
+        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
+        let dest_file = format!("{}/{}", &dest_src, &fid_path);
+        let mut dest_dir = PathBuf::from(&dest_file);
+
+        dest_dir.pop();
+        if dest_dir.is_dir() == false {
+            fs::create_dir_all(&dest_dir).await?;
+        }
+
+        let md = fs::metadata(&src_file).await?;
+        /* invoking mpifileutils is slow, so only do it for directories
+         * and big files, otherwise just use rsync
+         */
+        if md.is_dir() || (md.len() > 1024 * 1024 * 1024) {
+            let output = Command::new("/usr/lib64/openmpi/bin/mpirun")
+                .arg("--hostfile")
+                .arg("/etc/iml/filesync-hostfile")
+                .arg("--allow-run-as-root")
+                .arg("dsync")
+                .arg("-S")
+                .arg(src_file)
+                .arg(dest_file)
+                .output();
+
+            let output = output.await?;
+            result.push(FidError {
+                fid: fid.fid.clone(),
+                data: fid.data.clone(),
+                errno: output.status.code().unwrap_or(0) as i16,
+            });
+            tracing::debug!(
+                "dsync exited with {} {} {}",
+                output.status,
+                std::str::from_utf8(&output.stdout)?,
+                std::str::from_utf8(&output.stderr)?
+            );
+        } else {
+            rsync_list.push(Work {
+                src_file: src_file.clone(),
+                dest_file: dest_file.clone(),
+            });
+
+            if rsync_list.len() >= 500 {
+                let mut rsync_res = do_rsync(&rsync_list).await?;
+
+                result.append(&mut rsync_res);
+                rsync_res.clear();
+                rsync_list.clear();
+            }
+        }
+    }
+
+    let mut rsync_res = do_rsync(&rsync_list).await?;
```

Comment:
```suggestion
   do_rsync(&rsync_list).await;
```

The way do_rsync is currently constructed, it neither populates results or propagates errors.

Reply:
Fixed (reconstructed)